### PR TITLE
Event: evaluate selector at add time

### PR DESCRIPTION
**Pull request by fsateler**
### Summary ###

This ensures events with invalid selectors throw at attach time instead
of event time. This is done by instantiating a Sizzle for the selector.

`jQuery.find()` is used instead of `jQuery.find.matchesSelector` (as suggested in https://github.com/jquery/jquery/issues/3071#issuecomment-213254660) , because `matchesSelector` requires a `DOMElement`, which won't be if the element is actually document.

Fixes: #3071 

### Checklist ###
Mark an `[x]` for completed items, if you're not sure leave them unchecked and we can assist.

* [x] All authors have signed the CLA at https://contribute.jquery.com/CLA/
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [ ] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

Thanks! Bots and humans will be around shortly to check it out.

### CODE REVIEW

File: src/event.js [hunk 1]
```diff
@@ -120,6 +120,11 @@ jQuery.event = {
 			selector = handleObjIn.selector;
 		}
 
+		// If the selector is invalid, throw any exceptions at attach time
+		if ( selector ) {
+			jQuery.find( selector, elem );
```
Comment (timmywil) [inline src/event.js]: This may or may not require a context `elem.ownerDocument`. If it's just testing validity, perhaps not, but can anyone think of a use case where the document would affect validity?

Comment (dmethvin) [inline src/event.js]: None that I can think of.  The only case I can think of that breaks is if a selector extension was added after the event was attached here, but before an event was fired. That's probably not common?

Comment (dmethvin) [inline src/event.js]: Also note that this is inside `jQuery.event.add()` and the API surface is `jQuery.fn.on()`. If someone does `$("tr").on("mouseover" ...)` on a table with 100 elements it will check the selector 100 times. The parsed selector is cached and the selection set is small obviously, but is there a downside to moving this up to [function on()](https://github.com/fsateler/jquery/blob/10dc9f515a97c69ac674cfa19298ddd1e5cecbba/src/event.js#L91)?

File: test/unit/event.js [hunk 2]
```diff
@@ -1289,6 +1289,27 @@ QUnit.test( "Delegated events in SVG (#10791; #13180)", function( assert ) {
...
 } );
 
+QUnit.test( "Delegated events with malformed selectors (#3071)", function( assert ) {
+	assert.expect( 2 );
+	try {
+		jQuery( "#qunit-fixture" ).on( "click", "div:not", function () { } );
+		assert.ok( false, "malformed selector throws on attach" );
+	}
+	catch ( e ) {
+		assert.ok( true, "malformed selector throws on attach" );
```
Comment (gibson042) [inline test/unit/event.js]: I think this would be a little clearer with [`assert.throws`](http://api.qunitjs.com/throws/), but otherwise :+1:.

File: src/event.js (hunk 1 above)
Comment (fsateler) [inline src/event.js]: The only problem I see is that one would have to replicate the selector parameter parsing. Maybe a better approach would be to check if there is already an event handler with such a selector in the `handlers` list?

I may be wrong, but I think event handlers with selectors usually attach to a single object, as the whole point is to have to avoid listing all possible elements (which may be a lot, or not exist yet).

File: test/unit/event.js (hunk 2 above)
Comment (fsateler) [inline test/unit/event.js]: That would leave this asymmetric with the below call (I don't see a `notThrows` assert method). I you think that doesn't matter, I can change that.

Comment (dmethvin) [inline test/unit/event.js]: If the test throws and you don't expect it to, you find out pretty quickly. :smile:

Comment (fsateler) [inline test/unit/event.js]: OK, so I just let qUnit handle the exception.

To fix these issues, should I force-push, or open a new PR ?

Comment (gibson042) [inline test/unit/event.js]: I'd recommend just adding one or more new commits and pushing... no `-f` required.

Comment (dmethvin) [inline test/unit/event.js]: Generally you can just add a new commit that says `squash! (comment about the change)` and we will squash it when the PR lands. That way the commits here in the PR and the comments show the history of how we got to the squashed PR.

File: src/event.js (hunk 1 above)
Comment (dmethvin) [inline src/event.js]: @fsateler You are right, it won't be common for people to make a call to attach delegated events on a lot of elements at once. Also, this change does have the potential to throw errors where they did not before, in the case where the selector was invalid but no event ever occurred. We learned in gh-2824 that even obscure cases like that can cause trouble. So even if we moved this to `on()` it should be executed only if there is at least one element.

In short, I think you have this check in a good place.

**Comment by dmethvin** [comment]
@fsateler can you make that `assert.throws` tweak to the tests? I think this is good to land otherwise.

Anyone have thoughts on whether and how this should be documented? I am thinking it doesn't need docs, this behavior may have been what everyone expected anyway.

**Comment by fsateler** [comment]
Yes, I've just been busy during this week with $work. I'll push the change during the weekend.

**Comment by fsateler** [comment]
Updated for the `throws` change.

**Comment by gibson042** [comment]
Thanks, @fsateler! As it turns out, `matchesSelector` was necessary after all.
